src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReferenceEqualsEqualityComparer.cs

```diff
@@ -0,0 +1,27 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+
+namespace System.Text.Json
+{
+    /// <summary>
+    /// Passed to the <see cref="DefaultReferenceResolver._objectKeyMap"/> meant for serialization.
+    /// It forces the dictionary to do a ReferenceEquals comparison when comparing the TKey object.
+    /// </summary>
+    internal class ReferenceEqualsEqualityComparer<T> : IEqualityComparer<T>
+    {
+        public static ReferenceEqualsEqualityComparer<T> Comparer = new ReferenceEqualsEqualityComparer<T>();
+
+        bool IEqualityComparer<T>.Equals(T x, T y)
+        {
+            return ReferenceEquals(x, y);
+        }
+
+        int IEqualityComparer<T>.GetHashCode(T obj)
+        {
+            return GetHashCode();
```

Comment:
Should this be the following?
```suggestion
            return obj.GetHashCode();
```

Reply:
You are right, I was wrongly returning the hash of the `EqualityComparer` instance instead of the `obj`.